modules/system/utility-wfrp4e.js

```diff
@@ -1080,9 +1080,30 @@ export default class WFRP_Utility {
         }
       }
     }
-    let asyncFunction = Object.getPrototypeOf(async function () { }).constructor
-    let func = new asyncFunction("args", getProperty(effect, "flags.wfrp4e.script")).bind({ actor, effect })
-    func({ actor })
+
+    WFRP_Utility.runSingleEffect(effect, actor, null, { actor });
+  }
+
+  static async runSingleEffect(effect, actor, item, scriptArgs) {
+    try {
+      let func;
+      if (effect.script?.indexOf("await") == -1) {
+        func = new Function("args", effect.script).bind({ actor, effect, item })
+        WFRP_Utility.log(`${this.name} > Running ${effect.label}`)
+      } else if (effect.script?.indexOf("await") != -1) {
+        let asyncFunction = Object.getPrototypeOf(async function () { }).constructor
+        func = new asyncFunction("args", effect.script).bind({ actor, effect, item })
+        WFRP_Utility.log(`${this.name} > Running Async ${effect.label}`)
+      }
+      if(func) {
+        func(scriptArgs);
+      }
+    }
+    catch (ex) {
+      ui.notifications.error(game.i18n.format("ERROR.EFFECT", { effect: effect.label }))
+      console.error("Error when running effect " + effect.label + " - If this effect comes from an official module, try replacing the actor/item from the one in the compendium. If it still throws this error, please use the Bug Reporter and paste the details below, as well as selecting which module and 'Effect Report' as the label.")
+      console.error(`REPORT\n-------------------\nEFFECT:\t${effect.label}\nACTOR:\t${actor.name} - ${actor.id}\nERROR:\t${ex}`)
+    }
```

Comment:
Why was this needed exactly? There are specific triggers that can't support asynchronous execution and specific ones that can. The specific ones were given async functions already.

Reply:
first of all, to unify the logging errors and how effects are executed. Second of all, to allow async/await in all effects, because its not documented anywhere why and which effects may be async and which not. And even if - i dont get why only certain effects can be async? you are not awaiting any method in runEffects so it doesnt matter. In fact, you might want to look into my experimental branch https://github.com/silentmark/WFRP4e-FoundryVTT/tree/features/experimental where im trying to await all functiosn that are async, including effects.

Reply:
Any effect that gets called during actor preparation for example. Or really any time the calling function isn't async, calling effects that are won't really work / will be unpredictable.

Reply:
While I quite like unifying the code here, something about searching for "await" rubs me the wrong way, can't we still just use the `async` flag?

Reply:
If you look at the current implementation, `async` is just an option passed
```js
        if (!options.async)
          func = new Function("args", e.script).bind({ actor: this, effect: e, item: e.item })
        else if (options.async) {
          let asyncFunction = Object.getPrototypeOf(async function () { }).constructor
          func = new asyncFunction("args", e.script).bind({ actor: this, effect: e, item: e.item })
        }
```

It is underutilized however, I generally add it to a trigger when I need it to work for a specific effect and if the trigger supports it. Using callbacks was from before I added async support to effect scripts at all and tend to not do that anymore, only when I'm copy-pasting or it requires it.

`turn.actor.runEffects("endRound", combat, {async: true})`

As for what supports them, what absolutely cannot are preparation ones, such as `(pre)prepareData`, `(pre)prepareItem(s)`, `preWoundCalc`, etc. A comprehensive list would be a good thing to add to a hypothetical effects tutorial

As for why, an example would be an async `preWoundCalc` changes `sb`, `tb`, or any other of its arguments, won't happen in time for the calculation.

What could possibly support them would be `prefillDialog`? The calling function isn't async, so I think it would be unpredictable, and should probably change it to `async` first.

Point is, the `async` flag should be used and only on the ones that can fully support it. 

I would ask to hold off changing existing triggers (like `prefillDialog`) to support async for a separate PR though

Reply:
sure, i will change it, but i will leave the unified function. Once i will have more time to test my experimental branch, i will create a PR that should allow to use async functions in more broad way, and also most of them should be properly awaited. I think that this lack of awaiting asynchronous functions sometimes leads to those anoying errors with CreateEmbeddedDocument or DeleteEmbeddedDocument.